lib/Runtime/Types/PathTypeHandler.cpp

```diff
@@ -809,19 +857,138 @@ namespace Js
+
+        PropertyIndex propertyIndex = GetTypePath()->LookupInline(propertyId, GetPathLength());
+        if (propertyIndex == Constants::NoSlot)
+        {
+            PropertyRecord const* propertyRecord = scriptContext->GetPropertyName(propertyId);
+            if (propertyRecord->IsNumeric())
+            {
+                return ConvertToDictionaryType(instance)->SetItemAccessors(instance, propertyRecord->GetNumericValue(), getter, setter);
+            }
+
+            // We'll add 2 properties to the type, so check the limit.
+            if (GetPathLength() + 2 > TypePath::MaxPathTypeHandlerLength)
+            {
+                return ConvertToDictionaryType(instance)->SetAccessors(instance, propertyId, getter, setter, flags);
+            }
+
+            getter = CanonicalizeAccessor(getter, library);
+            setter = CanonicalizeAccessor(setter, library);
+
+            propertyIndex = GetPathLength();
+            AddPropertyInternal(instance, propertyId, getter, (ObjectSlotAttributes)(ObjectSlotAttr_Default | ObjectSlotAttr_Accessor), nullptr, flags, SideEffects_None);
+            newTypeHandler = PathTypeHandlerBase::FromTypeHandler(instance->GetTypeHandler());
+            Assert(propertyIndex == newTypeHandler->GetTypePath()->LookupInline(propertyId, newTypeHandler->GetPathLength()));
+        }
+        else
+        {
+            ObjectSlotAttributes attr = attributes ? attributes[propertyIndex] : ObjectSlotAttr_Default;
+            if (!(attr & ObjectSlotAttr_Accessor))
+            {
+                getter = CanonicalizeAccessor(getter, library);
+                setter = CanonicalizeAccessor(setter, library);
+  
+                if (!setters || setters[propertyIndex] == NoSetterSlot)
+                {
+                    // We'll add 1 property to the type, so check the limit.
+                    if (GetPathLength() + 1 > TypePath::MaxPathTypeHandlerLength)
+                    {
+                        return ConvertToDictionaryType(instance)->SetAccessors(instance, propertyId, getter, setter, flags);
+                    }
+                }
+
+                SetAttributesHelper(instance, propertyId, propertyIndex, attributes, (ObjectSlotAttributes)(attr | ObjectSlotAttr_Writable | ObjectSlotAttr_Accessor));
+                // SetAttributesHelper can convert to dictionary in corner cases, e.g., if we haven't got a full path from the root. Remove this check when that's fixed.
+                if (!instance->GetTypeHandler()->IsPathTypeHandler())
+                {
+                    return instance->GetTypeHandler()->SetAccessors(instance, propertyId, getter, setter, flags);
+                }
+                newTypeHandler = PathTypeHandlerBase::FromTypeHandler(instance->GetTypeHandler());
+            }
+            if (getter != nullptr)
+            {
+                newTypeHandler->SetSlotUnchecked(instance, propertyIndex, CanonicalizeAccessor(getter, library));
```

Comment:
>SetSlotUnchecked [](start = 32, length = 16)

Would we possibly want to call SetSlotAndCache instead of SetSlotUnchecked?

Reply:
We're not currently set up to use PropertyValueInfo on this path, so we can't use SetSlotAndCache. Currently DictionaryTypeHandler::SetAccessors is the only path for setting/adding accessors to a type handler, and it calls SetSlotUnchecked.

Reply:
BTW, we're setting accessor inline caches at LdFld/StFld that call the accessors, so we're covered. We enter this method at defineProperty/object literal/etc., so setting an inline cache here isn't probably appropriate.

Reply:
Agreed. But AddPropertyInternal in the 'if' clause above ends up calling SetSlotAndCache.

---
In reply to: [169468872](https://github.com/Microsoft/ChakraCore/pull/4283#discussion_r169468872) [](ancestors = 169468872)

Reply:
Discussed this offline. The gist is that in a non-add-property case the cache only needs to be updated when the type is changing, under SetAttributesHelper. We should investigate what we can do to make that path update the cache.